bgpd/bgp_packet.c

```diff
@@ -1319,11 +1319,18 @@ static int bgp_open_receive(struct peer *peer, bgp_size_t size)
 	else
 		peer->v_holdtime = send_holdtime;
 
-	if ((CHECK_FLAG(peer->flags, PEER_FLAG_TIMER))
-	    && (peer->keepalive < peer->v_holdtime / 3))
-		peer->v_keepalive = peer->keepalive;
-	else
-		peer->v_keepalive = peer->v_holdtime / 3;
+	/* Set effective keepalive to 1/3 the effective holdtime.
+	 * Use configured keeplive when < effective keepalive.
+	 */
+	peer->v_keepalive = peer->v_holdtime / 3;
+	if (CHECK_FLAG(peer->flags, PEER_FLAG_TIMER)) {
+		if ((peer->keepalive) && (peer->keepalive < peer->v_keepalive))
+			peer->v_keepalive = peer->keepalive;
+	} else {
+		if ((peer->bgp->default_keepalive)
```

Comment:
I guess `peer->bgp->default_keepalive` is valid always since it has a default value on creation. Do we need this test still?

Reply:
I use this test here to make sure the value is non-0, since we still allow a user to configure `timers bgp 0 <non-zero>`. In that case, we don't want to use a default_keepalive of 0 when the holdtime is non-zero because it causes BGP flaps every $holdtime seconds.

For the 0/0 case, the holdtime negotiation would settle on an effective value of 0 then the effective keepalive is initially set to 0/3 (so also 0), which is the correct behavior when holdtime=0.  Since any peer/default value is not <0, we continue using 0 as the effective keepalive and we're g2g.

Reply:
got it